Unfinished pull

Fl4m3Ph03n1x opened the pull request:
Here. 

I was super original and named it "unfinished pull". 
It passes all the tape tests, for what it's worth. 
The bench script also runs. 
Updated several vulnerable dependencies. 
No more security issues with Buffer. 
Needs Node v6 or higher.

I could use a hand in testing it with a real system though, as I didn't have enough time for it yet, but all seems to be well.

The merge was quite hard to be honest. But anyway, what do you think?

Commit:
Added benchmark script

Commit:
Added benchmark script

Commit:
Rewrote the library for clarity and performance

Commit:
Merge branch 'bench' into refactor2

Commit:
Enhanced benchmark

Commit:
Refactoring 2

Commit:
Unit tests and virtually full coverage

Commit:
updated-readme

Fl4m3Ph03n1x committed:
resolved merged conflict. Replaced use of new Buffer with recommended Buffer usage.

Fl4m3Ph03n1x committed:
updated packages for vulnerability fixes

Fl4m3Ph03n1x committed:
merging with PR

Fl4m3Ph03n1x committed:
removed unused library

Fl4m3Ph03n1x committed:
remove useless files

Fl4m3Ph03n1x committed:
replaced use of new Buffer for security reasons

Fl4m3Ph03n1x committed:
updated packages to fix vulnerabilities

Fl4m3Ph03n1x committed:
removed deprecated usage of 'module.exports.graylog'

Fl4m3Ph03n1x committed:
moved exports to the end. Where it should be.

Fl4m3Ph03n1x committed:
fixed bench script to run without deprecated mode.

Fl4m3Ph03n1x committed:
updated vulnerable deps

ronkorving commented on package.json:
```diff
@@ -17,19 +21,27 @@
+  },
+  "dependencies": {
+    "jshint": "^2.9.5",
+    "nyc": "^12.0.2"
```
These are dev deps :)

Fl4m3Ph03n1x replied:
Yes they are, I wonder what they are doing there. 
I will fix that and try again!

ronkorving commented on lib/Graylog.js:
```diff
@@ -424,4 +424,3 @@ Graylog.prototype.close = function (cb) {
 
 
 module.exports = Graylog;
-module.exports.graylog = Graylog; // deprecated
```
This one line keeps API compatibility between the old and the new. That's why I added it to the re-implementation.

Fl4m3Ph03n1x replied:
Is it really necessary though? 
I did have the idea that this was supposed to be a major version bump, which means it won't be backwards compatible.

Fl4m3Ph03n1x committed:
fixed dependencies missing

Fl4m3Ph03n1x committed:
updating changelog

Fl4m3Ph03n1x commented:
moved dependencies to DEV, updated changelog with some preliminar information ( will need to confirm and change as we see what has actually be done and what has not ). 

For me, personally, I would make the log events async ( via Promises ) and get rid of the need to close the Graylog object for it to empty the queue. This need to close the socket is also not documented at all and I argue that we should ditch it altogether and just make the logs async. 

Since this is supposed to be a major version bump, why not get rid of all the issues we currently have?

Fl4m3Ph03n1x commented:
Hello there, do you have any news on this? I would like to have some more feedback to know what I can change so this can be approved.

ronkorving commented:
A lot of indentation got wonky. Please have a look at `Graylog.js`.
